modules/database/src/std/filters/ts.c

```diff
@@ -9,27 +10,130 @@
 
 /*
  *  Author: Ralph Lange <Ralph.Lange@bessy.de>
+ *  Author: Jure Varlec <jure.varlec@cosylab.com>
  */
 
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 
 #include "chfPlugin.h"
-#include "db_field_log.h"
 #include "dbExtractArray.h"
 #include "dbLock.h"
+#include "db_field_log.h"
 #include "epicsExport.h"
+#include "freeList.h"
+#include "cantProceed.h"
 
-/*
- * The size of the data is different for each channel, and can even
- * change at runtime, so a freeList doesn't make much sense here.
- */
+/* Allocation size for freelists */
+#define ALLOC_NUM_ELEMENTS 32
+
+
+/* Filter settings */
+
+enum tsMode {
+    tsModeInvalid = 0,
+    tsModeGenerate = 1,
+    tsModeDouble = 2,
+    tsModeSec = 3,
+    tsModeNsec = 4,
+    tsModeArray = 5,
+    tsModeString = 6,
+};
+
+static const chfPluginEnumType ts_numeric_enum[] = {
+    {"dbl", 2}, {"sec", 3}, {"nsec", 4}, {"ts", 5}};
+
+enum tsEpoch {
+    tsEpochEpics = 0,
+    tsEpochUnix = 1,
+};
+
+static const chfPluginEnumType ts_epoch_enum[] = {{"epics", 0}, {"unix", 1}};
+
+enum tsString {
+    tsStringInvalid = 0,
+    tsStringEpics = 1,
+    tsStringIso = 2,
+};
+
+static const chfPluginEnumType ts_string_enum[] = {{"epics", 1}, {"iso", 2}};
+
+typedef struct tsPrivate {
+    enum tsMode mode;
+    enum tsEpoch epoch;
+    enum tsString str;
+} tsPrivate;
+
+static const chfPluginArgDef ts_args[] = {
+    chfEnum(tsPrivate, mode, "num", 0, 0, ts_numeric_enum),
+    chfEnum(tsPrivate, epoch, "epoch", 0, 0, ts_epoch_enum),
+    chfEnum(tsPrivate, str, "str", 0, 0, ts_string_enum),
+    chfPluginArgEnd
+};
+
+static int parse_finished(void *pvt) {
+    tsPrivate *settings = (tsPrivate *)pvt;
+    if (settings->str != tsStringInvalid) {
```

Comment:
Is this the condition which justifies that `cantProceed()` below?

Reply:
I'm not sure what you mean with "justifies". But let me explain. The possible states after parsing are:
- no arguments were given, mode is `tsModeInvalid`, and we change that to `tsModeGenerate`;
- "num" was given, mode is one of the numeric modes, nothing needs to be done here;
- "str" was given, so we set mode accordingly.

So in this sense, `cantProceed()` calls further down are justified because those situations shouldn't be possible, and I wanted the `switch` statements to be exhaustive.

Reply:
Core group 8 June: Failing with `cantProceed()` is not appropriate in code that can be triggered remotely by any client, that could be a DoS which we'd rather avoid, please handle this some other way. It is definitely appropriate to log a message in the event of a logic error though.

Reply:
I see your point. Even if there was a straightforward way to prove the code was unreachable, it may not remain that way when someone changes it. Fixed in [e23c4bc](https://github.com/epics-base/epics-base/pull/126/commits/e23c4bc91332ebdb10f52ab68081c25c4c6cdd96).